keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+
+#### Should we use feature gate for bug fixes?
+
+There are some use cases when some bugs are found and later fixed in a patch version. The client would need to know if the etcd version in use contains that fix to decide whether or not to use that feature. 
+
+The question is: should a new feature gate be added to signal the bug fix? 
+
+We think the answer is generally NO:
+* the new feature would need to be enabled by default to always apply the bug fix for new releases.
+* it changes the API which is not desirable in patch version releases.
+
+The proper way of handling these bug fixes should be:
+1. the feature should be gated by the feature gate from the beginning.
+1. the feature should be disabled by default until it is widely tested in practice. 
+1. when the bug is found, the feature should ideally be at a lifecycle in which it is disabled by default. If not, the admin should disable it by the `--feature-gates` flag.
+1. when the client upgrades etcd to the patch version with the fix, the admin could enable it by the `--feature-gates` flag.
+
+But if the bug fix is very critical, it could be discussed in a case-by-case manner.
+
+#### Could the lifecycle of a feature change in patch versions?
+
+Kubernetes have a minor release every 3 months, while the cadence of etcd minor releases is much less frequent. The question is: do we have to wait for years before graduating a new feature?
+
+We think we should still stick to the common practice of not changing the lifecycle of a feature in patch versions. Because:
+* changing the lifecycle of a feature is an API change. According to the [etcd Operations Guide](https://etcd.io/docs/v3.5/op-guide/versioning/), only new minor versions may add additional features to the API.
+* bugs in etcd could be hard to detect, and the reliability and robustness of etcd is more important than speed. A long history of testing through practical adoption a new feature is beneficial.
+
+With the feature gate in place, we could consider increasing the etcd release cadence because it would be easier to add new features and less risky to release new features.
+
+### Risks and Mitigations
+
+* As discussed before, we are deferring cluster level feature support to a follow-up KEP.
+
+## Design Details
+
+### New APIs
+
+A new `--feature-gates` command line argument would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.
+
+New Prometheus gauge metrics would be added to monitor and query if a feature is enabled for the server. 
+For easier usage in Kubernetes, we will also add a grpc backed `clientv3` function and a new `etcdctl` command.
+
+The metrics could look like:
+* `etcd_feature_enabled{name="featureName",stage="Alpha"} 1` if the feature is enabled.
+
+For grpc, new rpc type would be added to the `Maintenance` service.
```

Comment:
For a server level feature I don't see any benefits of having dedicated API. Kubernetes doesn't have dedicated API.

I think there are considerations regarding if client can depend on feature flags not changing that would require a deeper discussion that would need to include cluster level feature flags. Please skip the API changes for now.

Reply:
removed.